mtags/src/main/scala-3/scala/meta/internal/pc/completions/MatchCaseCompletions.scala

```diff
@@ -0,0 +1,418 @@
+        autoImport match
+          case Some(value) =>
+            visit(sym.info.dealias.typeSymbol, sym.decodedName, value)
+          case scala.None =>
+            visit(sym.info.dealias.typeSymbol, sym.showFullName, Nil)
+      else ()
+    }
+
+    // Step 3: special handle case when selector is a tuple or `FunctionN`.
+    if definitions.isTupleClass(selectorSym) || definitions.isFunctionClass(
+        selectorSym
+      )
+    then
+      val label =
+        if !patternOnly then s"case ${parents.selector.show} =>"
+        else parents.selector.show
+      result += CompletionValue.CaseKeyword(
+        selectorSym,
+        label,
+        Some(
+          if !patternOnly then
+            if config.isCompletionSnippetsEnabled() then "case ($0) =>"
+            else "case () =>"
+          else if config.isCompletionSnippetsEnabled() then "($0)"
+          else "()"
+        ),
+        Nil,
+        range = Some(completionPos.toEditRange),
+        command = config.parameterHintsCommand().asScala,
+      )
+    end if
+
+    val res = result.result()
+    res
+  end contribute
+
+  /**
+   * A `match` keyword completion to generate an exhaustive pattern match for sealed types.
+   *
+   * @param selector the match expression being deconstructed or `EmptyTree` when
+   *                 not in a match expression (for example `List(1).foreach { case@@ }`.
+   * @param completionPos the position of the completion
+   * @param typedtree typed tree of the file, used for generating auto imports
+   * @param indexedContext
+   * @param config
+   */
+  def matchContribute(
+      selector: Tree,
+      completionPos: CompletionPos,
+      typedTree: Tree,
+      indexedContext: IndexedContext,
+      config: PresentationCompilerConfig,
+  ): List[CompletionValue] =
+    import indexedContext.ctx
+    val pos = completionPos.sourcePos
+    val definitions = indexedContext.ctx.definitions
+    val text = pos.source.content().mkString
+    val clientSupportsSnippets = config.isCompletionSnippetsEnabled()
+    lazy val autoImportsGen = AutoImports.generator(
+      pos,
+      text,
+      typedTree,
+      indexedContext,
+      config,
+    )
+    val completionGenerator = CompletionValueGenerator(
+      indexedContext,
+      completionPos,
+      clientSupportsSnippets,
+    )
+    val result = ListBuffer.empty[CompletionValue]
+    val tpe = selector.tpe.widen.bounds.hi match
+      case tr @ TypeRef(_, _) => tr.underlying
+      case t => t
+
+    def subclassesForType(tpe: Type)(using Context): List[Symbol] =
+      tpe.typeSymbol.info
+      val subclasses = ListBuffer.empty[Symbol]
+      val parents = ListBuffer.empty[Symbol]
+      // Get parent types from refined type
+      def loop(tpe: Type): Unit =
+        tpe match
+          case AndType(tp1, tp2) =>
+            loop(tp1)
+            loop(tp2)
+          case t => parents += tpe.typeSymbol
+      loop(tpe.widen.bounds.hi)
```

Comment:
What are we doing here? Is there any test that reflects these lines?

Reply:
Yes, this one https://github.com/scalameta/metals/blob/0443f440379712314c85a2b77a6707177cb326a0/tests/cross/src/test/scala/tests/pc/CompletionMatchSuite.scala#L274-L275
I will add comments explaining it